Add rerun option for workflows

ahmedhamidawan opened the pull request:
Fixes https://github.com/galaxyproject/galaxy/issues/11340

https://github.com/user-attachments/assets/eb05c16c-60f7-47a6-85c2-57d51beea2f9

<details>
<summary><h3>Earlier implementation</h3></summary>

https://github.com/user-attachments/assets/a0df13b4-9f7e-4853-b4e0-9e8a7d87d63e

</details>

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
add direct route to `WorkflowRun`

ahmedhamidawan committed:
allow workflow instances to be run via route prop

ahmedhamidawan committed:
Add rerun option for invocations

Fixes https://github.com/galaxyproject/galaxy/issues/11340

ahmedhamidawan committed:
change run option in `GridInvocation` to rerun as well

ahmedhamidawan committed:
use redo icon instead of play for workflow rerun

ahmedhamidawan committed:
inform user if rerun values are changed

We compare the `requestState` and the current values to tell whether the values are different from the original run at a given point. This check can be disabled.

Example case is if a dataset that was formerly used in an invocation, is now deleted, and therefore on landing on the rerun, is not set as the input.

ahmedhamidawan committed:
use more `GButton` in workflow title bar and fix recently replaced ones

ahmedhamidawan committed:
remove alert for disabling flagging changed inputs and move it to badge

Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>

ahmedhamidawan committed:
add a specific `isRerun` prop to make these changes specific to reruns

This separates the implementation of setting the `requestState` for workflow landings and workflow reruns. Also adds indication on the workflow execute button on whether a rerun will be done or not.

ahmedhamidawan committed:
add selenium test for workflow rerun

mvdbeek reviewed (approved):
Looks great, thanks a lot! I'd merge if you take it out of draft.

mvdbeek commented on lib/galaxy_test/selenium/test_workflow_rerun.py:
```diff
@@ -0,0 +1,89 @@
+        self.components.invocations.workflow_rerun_button.wait_for_and_click()
+        self.sleep_for(self.wait_types.UX_TRANSITION)
+
+        # Arriving on the rerun page, the history should have switched back to the original history
```
Oh, we should not switch the history. What's the reason for that ?
That seems rather surprising, and also not what we do for job reruns.

ahmedhamidawan replied:
We switch to the history because the workflow run form loads input datasets and collections from the current history.

Therefore, if the request state contains values from another history, we need to switch that history, so that the FormData components have the original inputs.

mvdbeek replied:
What is the reason for that ? We should not cement a bug with tests.

ahmedhamidawan replied:
I believe that's just how the run form has always been, right? We run a workflow with inputs from the current history

mvdbeek replied:
Jobs can be rerun from a different history. That uses the same form data component, why does that not work here ?

ahmedhamidawan replied:
Thats true. 

Let me rework how the workflow run form handles this then...

mvdbeek replied:
Well, that might sprawl (I assume, maybe it's easy ?). How about we issue a confirmation dialog if the history has to be switched for now ? Eventually I think we should be able to switch the current history without affecting the values selected in the run form.

ahmedhamidawan replied:
Sure, would it be fine to instead of opening a modal, do it right where the form would appear in the center?

Because to make this work, I made it load the history FIRST, then render the form.

ahmedhamidawan replied:
> Eventually I think we should be able to switch the current history without affecting the values selected in the run form.

I think that should definitely be the thing in the future.

It's not obvious to the average user that changing the history will make them lose all of their formdata values

mvdbeek replied:
> I made it load the history FIRST, then render the form.

What I was thinking is that we pop open the confirmation dialog as you click on rerun ? Since at that point you already know if you need to switch the history or not to do a rerun

ahmedhamidawan replied:
Oh that can work too yes, on cancel you stay at the invocation view, on confirm, you route to rerun (where we change history).

ahmedhamidawan replied:
Does this work?

There is one thing though. In this case, if the rerun history is different, and the user does not want to change their current history, all they can do is just cancel. What if they want to run the same workflow here but for a different history? 
To allow that I could use a custom modal instead of the confirm dialog (which only has confirm/cancel) with a 3rd option of just run workflow (instead of rerun with original inputs).

https://github.com/user-attachments/assets/af39b71a-9a72-4c92-807c-919eedc6a932

mvdbeek replied:
Yeah, three options or split the button into "Run workflow" and "Rerun workflow" ? That seems fine.

ahmedhamidawan replied:
I thought about Run | Rerun too
I think I'll do that

ahmedhamidawan replied:
![image](https://github.com/user-attachments/assets/b6e26138-e60f-44cc-9396-41df4a338708)

Pushed this change.

ahmedhamidawan committed:
add confirmation dialog to let user know history will change on rerun

ahmedhamidawan committed:
handle "" keys values from GET invocations/{id}/request in rerun

Sometimes, `/api/invocations/{invocation_id}/request` can return "" key values for inputs and that would cause incorrect values to be set as expected rerun values due to a key mismatch.

ahmedhamidawan committed:
add run option next to rerun if users just want to run the same workflow

ahmedhamidawan committed:
adjust selenium for added confirm dialog for rerun

ahmedhamidawan committed:
use `.g-disabled` class for disabled workflow run selector
